### ISSUE TITLE
Have `set --show` give location of config file that is setting the global variable

### ISSUE THREAD

**Issue by devansh08** [fish-shell/fish-shell#8822]
I'm running fish version `3.3.1`.

So this is like a feature request and an issue I'm currently facing and need help with.

My `JAVA_HOME` env variable being set with global context from some config that I am not able to find. I want to update its value and use it in universal context but even if i erase it and set with `-U`, in a new session the global value is back and is preferred (assuming this is expected ?).

What I hoped for is `set --show JAVA_HOME` to show what is actually setting this variable on starting a new session, as I understand universal variables are handled via tmp files and local variables are not set when session starts. So there is for sure some config that is setting this global variable.

If it helps, this is what `set` shows me:
```shell
set --show JAVA_HOME
$JAVA_HOME: set in global scope, exported, with 1 elements
$JAVA_HOME[1]: |/usr/lib/jvm/java-15-jdk|
$JAVA_HOME: set in universal scope, exported, with 1 elements
$JAVA_HOME[1]: |/usr/lib/jvm/default|
```

I've also tried checking this variable as `root` user and also with a temp home directory (`sh -c 'env HOME=$(mktemp -d) fish'`). In both cases this global variable is set (universal variable does not come up with `set --show`). 

Is there some way (via `set` or anything else) that can tell me where this variable is set ? Or a more crude way, if I can get the list of all config files fish reads/executes on session start ? I can go through those and see. So far I have checked `~/.config/fish/*`, `/root/.config/fish/*` and `/usr/share/fish/config`. Neither of these has it.

Any ideas/thoughts ?

**Comment by faho** [comment]
>I want to update its value and use it in universal context

You don't want to use it as a universal variable - see https://fishshell.com/docs/current/faq.html#why-doesn-t-set-ux-exported-universal-variables-seem-to-work.

>as I understand universal variables are handled via tmp files

This isn't *really* correct. Universal variables are written to *a specific file* on disk, and read from *that specific file*, really early.

>So there is for sure some config that is setting this global variable.

Quite possibly not - the variable is *exported*, which means it will be passed down to child processes, and they will inherit it, as a global variable - see the FAQ entry above.

If there is a fish somewhere above your terminal in the stack, like a login shell or something, or if some other process defined it, then you would be getting this effect, that the universal variable isn't useful.

(yes, this is a reason we're looking into replacing universal variables)

----

In short: What you want is to simply set this as a global-and-exported variable.

Delete the universal variable - `set -eU JAVA_HOME`, add a `set -gx JAVA_HOME whatever` to your config.fish and reboot.

**Comment by faho** [comment]
Duplicate of #1226

**Comment by devansh08** [comment]
> If there is a fish somewhere above your terminal in the stack, like a login shell or something, or if some other process defined it, then you would be getting this effect, that the universal variable isn't useful.
Ah... I did not know about this behaviour. I have `zsh` as my user default shell and `fish` is default for my terminal app. And my zsh config is setting the variable to the value I do not want. That is my issue here.

Reading the FAQ you linked helped me understand this. It makes sense now why universal variables would likely be a cause for such discrepancies... but they are just sooo convenient to use :)

> Universal variables are written to a specific file on disk, and read from that specific file, really early.
Out of curiosity, are these `~/.config/fish/fishd.tmp.*` the files you mean here ? I have like 750+ of them in my directory and majority are just empty.

**Comment by krobelus** [comment]
On Sun, Mar 20, 2022 at 03:02:18PM -0700, Devansh Sharma wrote:
> > Universal variables are written to a specific file on disk, and read from that specific file, really early.
> Out of curiosity, are these `~/.config/fish/fishd.tmp.*` the files you mean here ? I have like 750+ of them in my directory and majority are just empty.

yes except ~/.config/fish/fishd.* were used in old versions of fish. The replacement is ~/.config/fish/fish_variables, so you can delete the fishd.* files

**Comment by devansh08** [comment]
Oh cool ok. I did delete all expect the biggest one. But will go ahead and remove that too. Thanks.
